File: 백준/Gold/32177. 에어드롭/에어드롭.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        int t = Integer.parseInt(st.nextToken());

        int[][] locations = new int[n+1][3];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < 3; i++)
            locations[0][i] = Integer.parseInt(st.nextToken());

        boolean[] isPicture = new boolean[n+1];
        for (int i = 1; i <= n; i++) {
            st = new StringTokenizer(br.readLine());

            for (int j = 0; j < 3; j++)
                locations[i][j] = Integer.parseInt(st.nextToken());
            isPicture[i] = st.nextToken().equals("1");
        }
        
        List<Integer>[] graph = new ArrayList[n+1];
        for (int i = 0; i <= n; i++)
            graph[i] = new ArrayList<>();

        for (int i = 0; i <= n; i++) {
            for (int j = i+1; j <= n; j++) {
                if (canDelivery(locations, i, j, k, t)) {
                    graph[i].add(j);
                    graph[j].add(i);
                }
            }
        }

        List<Integer> answer = bfs(graph, isPicture, n);
        if (answer.isEmpty())
            System.out.println(0);
        else {
            StringBuilder sb = new StringBuilder();
            answer.sort(Comparator.comparingInt(i -> i));
            for (int x: answer)
                sb.append(x).append(" ");

            System.out.println(sb);
        }
    }

    private static boolean canDelivery(int[][] locations, int i, int j, int k, int t) {
        return Math.abs(locations[i][2] - locations[j][2]) <= t &&
                (locations[i][0] - locations[j][0]) * (locations[i][0] - locations[j][0]) + (locations[i][1] - locations[j][1]) * (locations[i][1] - locations[j][1]) <= k * k;
    }

    private static List<Integer> bfs(List<Integer>[] graph, boolean[] isPicture, int n) {
        Deque<Integer> q = new ArrayDeque<>();
        q.add(0);

        boolean[] visited = new boolean[n+1];
        visited[0] = true;

        List<Integer> picture = new ArrayList<>();
        while (!q.isEmpty()) {
            int now = q.poll();
            if (isPicture[now])
                picture.add(now);

            for (int net : graph[now]) {
                if (!visited[net]) {
                    visited[net] = true;
                    q.add(net);
                }
            }
        }

        return picture;
    }
}

File: 백준/Gold/32177. 에어드롭/에어드롭.py
import sys
from collections import deque
input = sys.stdin.readline


def can_deliver(x, y):
    xx, xy, xv = locations[x]
    yx, yy, yv = locations[y]

    return True if abs(xv-yv) <= t and (xx-yx) ** 2 + (xy-yy) ** 2 <= k*k else False


def bfs():
    q = deque()
    q.append(0)

    visited = [False] * (n+1)
    visited[0] = True

    pictures = []
    while q:
        now = q.popleft()
        if is_picture[now]:
            pictures.append(now)

        for nxt in graph[now]:
            if not visited[nxt]:
                visited[nxt] = True
                q.append(nxt)

    return pictures


n, k, t = map(int, input().split())

locations = []
is_picture = [0]
locations.append(list(map(int, input().split())))
for _ in range(n):
    *location, p = map(int, input().split())
    locations.append(location)
    is_picture.append(p)

graph = [[] for _ in range(n+1)]
for i in range(n+1):
    for j in range(i+1, n+1):
        if can_deliver(i, j):
            graph[i].append(j)
            graph[j].append(i)

answer = bfs()
print(' '.join(map(str, sorted(answer))) if answer else 0)